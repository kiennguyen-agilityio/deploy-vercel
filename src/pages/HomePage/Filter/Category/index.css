.subcategory-group {
  display: flex;
  flex-direction: column;
}

.filter-title:hover {
  background-color: var(--color-hover);
}

.dropdown-text:hover {
  color: var(--color-button-primary);
}

.price-button:hover {
  background-color: var(--color-background-secondary);
}

.subcategory:hover {
  background-color: var(--color-background-secondary);
}
