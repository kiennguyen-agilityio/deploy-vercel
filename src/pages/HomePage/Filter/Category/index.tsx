import { useState } from 'react';

// components
import FilterTitle from '@/pages/HomePage/Filter/Title';
import CategoryList from './CategoryList';
import SubCategoryList from './SubCategoryList';

// styles
import './index.css';

interface Props {
  selectedCategory: string;
  categoryOptions: string[];
  selectedSubCategory: string;
  subCategoryOptions: Record<string, string[]>;
  onSelectCategory: (category: string) => void;
  onSelectSubCategory: (subcategory: string) => void;
}

const CategoryFilter: React.FC<Props> = ({
  selectedCategory,
  categoryOptions,
  subCategoryOptions,
  selectedSubCategory,
  onSelectCategory,
  onSelectSubCategory,
}) => {
  const [isCategoryOpen, setIsCategoryOpen] = useState<boolean>(false);

  const handleDropdownOpen = (value: boolean): void => {
    setIsCategoryOpen(value);
  };

  return (
    <div className="filter-section">
      <FilterTitle onDropdownOpen={handleDropdownOpen} title="Category" icon="expand_more" />

      {isCategoryOpen && (
        <div className="dropdown-content">
          {categoryOptions.map((category) => (
            <div className="subcategory-group" key={category}>
              <CategoryList
                category={category}
                selectedCategory={selectedCategory}
                onSelectCategory={onSelectCategory}
              />
              {selectedCategory === category && (
                <div className="subcategory-list">
                  {subCategoryOptions[category].map((subcategory) => (
                    <SubCategoryList
                      key={subcategory}
                      subCategory={subcategory}
                      selectedSubCategory={selectedSubCategory}
                      onSelectSubCategory={onSelectSubCategory}
                      selectedCategory={selectedCategory}
                    />
                  ))}
                </div>
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
};

export default CategoryFilter;
