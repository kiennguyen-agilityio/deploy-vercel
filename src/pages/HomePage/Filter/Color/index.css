.dropdown-content {
  cursor: pointer;
  margin: 23px 10px 0 0;
}

.color-dropdown {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
}
