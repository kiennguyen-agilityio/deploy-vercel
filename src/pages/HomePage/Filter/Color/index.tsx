import { useState } from 'react';

// components
import Checkbox from '@/components/Checkbox';
import FilterTitle from '@/pages/HomePage/Filter/Title';

// styles
import './index.css';

interface Props {
  selectedColors: string[];
  colorOptions: string[];
  onColorChange: (colors: string[]) => void;
}

const ColorFilter: React.FC<Props> = ({ selectedColors, onColorChange, colorOptions }) => {
  const [isColorOpen, setIsColorOpen] = useState<boolean>(false);

  const handleColorChange = (event: React.ChangeEvent<HTMLInputElement>): void => {
    const color = event.target.value;
    if (selectedColors.includes(color)) {
      const updatedColors = selectedColors.filter((it) => it !== color);
      onColorChange(updatedColors);
    } else {
      const updatedColors = [...selectedColors, color];
      onColorChange(updatedColors);
    }
  };

  const handleDropdownOpen = (value: boolean): void => {
    setIsColorOpen(value);
  };

  return (
    <div className="filter-section">
      <FilterTitle onDropdownOpen={handleDropdownOpen} title="Color" icon="expand_more" />

      {isColorOpen && (
        <div className="dropdown-content">
          {colorOptions.map((color) => (
            <Checkbox
              key={color}
              id={color}
              value={color}
              checked={selectedColors.includes(color)}
              onChange={handleColorChange}
            />
          ))}
        </div>
      )}
    </div>
  );
};

export default ColorFilter;
