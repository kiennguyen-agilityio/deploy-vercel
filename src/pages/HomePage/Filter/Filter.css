
.filter-container {
  width: 245px;
  margin-left: 20px;
  display: inline-block;
}

.filter-header {
  margin-bottom: 15px;
  color: var(--color-border-secondary);
  font: normal var(--font-weight-bold) var(--font-size-sm) / var(--line-height-l) 'Open Sans';
}

.dropdown-text {
  margin-left: 15px;
}
