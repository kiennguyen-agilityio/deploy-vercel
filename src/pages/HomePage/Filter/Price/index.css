.price-button {
  display: flex;
  justify-content: normal;
  align-items: center;
  box-sizing: border-box;
  width: 235px;
  height: 57px;
  background: var(--color-background);
  border: 1px solid var(--color-border-tertiary);
  border-radius: 5px;
  padding: 0;
  margin-bottom: 30px;
}

.price-button-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 56px;
  height: 55px;
  background: var(--color-button-price);
  border-radius: 4px 0px 0px 4px;
}

.price-button-label {
  margin: 0 auto;
  width: 146px;
  height: 30px;
  font: normal var(--font-weight-regular) var(--font-size-sm) / var(--line-height-l) 'Open Sans';
  color: var(--color-primary);
}
