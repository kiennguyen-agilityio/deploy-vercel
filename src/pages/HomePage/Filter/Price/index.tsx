import { useState } from 'react';

// components
import FilterTitle from '@/pages/HomePage/Filter/Title';

// styles
import './index.css';

interface Props {
  onSortByPriceAsc: () => void;
  onSortByPriceDesc: () => void;
}

const PriceFilter: React.FC<Props> = ({ onSortByPriceAsc, onSortByPriceDesc }) => {
  const [isPriceOpen, setIsPriceOpen] = useState<boolean>(false);

  const handleDropdownOpen = (value: boolean): void => {
    setIsPriceOpen(value);
  };

  return (
    <div className="filter-price">
      <FilterTitle onDropdownOpen={handleDropdownOpen} title="Price" icon="expand_more" />
      {isPriceOpen && (
        <div className="dropdown-content">
          <button className="price-button" type="button" onClick={onSortByPriceDesc}>
            <div className="price-button-icon">$</div>
            <div className="price-button-label">Maximum Price</div>
          </button>
          <button className="price-button" type="button" onClick={onSortByPriceAsc}>
            <div className="price-button-icon">$</div>
            <div className="price-button-label">Minimum Price</div>
          </button>
        </div>
      )}
    </div>
  );
};

export default PriceFilter;
