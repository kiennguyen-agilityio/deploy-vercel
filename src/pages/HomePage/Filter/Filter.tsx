// helpers
import {
  colorOptions,
  ratingOptions,
  categoryOptions,
  subcategoryOptions,
} from '@/constants/options';

// components
import ColorFilter from '@/pages/HomePage/Filter/Color';
import CategoryFilter from '@/pages/HomePage/Filter/Category';
import RatingFilter from '@/pages/HomePage/Filter/Rating';
import PriceFilter from '@/pages/HomePage/Filter/Price';

// styles
import './Filter.css';

export interface Props {
  selectedColors: string[];
  selectedCategory: string;
  selectedSubcategory: string;
  onColorChange: (colors: string[]) => void;
  onCategoryChange: (category: string) => void;
  onSubcategoryChange: (subcategory: string) => void;
  onSortByPriceAsc: () => void;
  onSortByPriceDesc: () => void;
}

const Filter: React.FC<Props> = ({
  selectedColors,
  selectedCategory,
  selectedSubcategory,
  onColorChange,
  onCategoryChange,
  onSortByPriceAsc,
  onSortByPriceDesc,
  onSubcategoryChange,
}) => {
  const handleSortByPriceAsc = (): void => {
    onSortByPriceAsc();
  };

  const handleSortByPriceDesc = (): void => {
    onSortByPriceDesc();
  };

  const handleCategorySelect = (category: string): void => {
    onCategoryChange(category);
  };

  const handleSubcategorySelect = (subcategory: string): void => {
    onSubcategoryChange(subcategory);
  };

  return (
    <div className="filter-container">
      <h2 className="filter-header">Filter</h2>

      <ColorFilter
        selectedColors={selectedColors}
        onColorChange={onColorChange}
        colorOptions={colorOptions}
      />

      <RatingFilter ratingOptions={ratingOptions} />

      <PriceFilter
        onSortByPriceAsc={handleSortByPriceAsc}
        onSortByPriceDesc={handleSortByPriceDesc}
      />

      <CategoryFilter
        selectedCategory={selectedCategory}
        onSelectCategory={handleCategorySelect}
        categoryOptions={categoryOptions}
        subCategoryOptions={subcategoryOptions}
        selectedSubCategory={selectedSubcategory}
        onSelectSubCategory={handleSubcategorySelect}
      />
    </div>
  );
};

export default Filter;
