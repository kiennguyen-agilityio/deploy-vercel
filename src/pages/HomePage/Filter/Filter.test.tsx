import { render, fireEvent, screen } from '@testing-library/react';
import Filter, { Props } from './Filter';

const mockProps = {
  selectedColors: [],
  selectedCategory: '',
  onColorChange: jest.fn(),
  onCategoryChange: jest.fn(),
  onSubcategoryChange: jest.fn(),
  onSortByPriceAsc: jest.fn(),
  onSortByPriceDesc: jest.fn(),
  selectedSubcategory: '',
  isColorOpen: false,
};

const setup = (overrides: Props) => {
  const props = { ...mockProps, ...overrides };

  return render(<Filter {...props} />);
};

describe('Filter component', () => {
  it('renders correctly', () => {
    const { asFragment } = setup(mockProps);
    expect(asFragment()).toMatchSnapshot();
  });
  it('calls onColorChange when a color checkbox is clicked', () => {
    const { getByTestId, getByLabelText } = render(<Filter {...mockProps} />);

    const expandMoreButton = getByTestId('expand-color');
    fireEvent.click(expandMoreButton);

    const checkbox = getByLabelText('Red');
    fireEvent.click(checkbox);

    expect(mockProps.onColorChange).toHaveBeenCalledWith(['Red']);
  });

  it('calls onSortByPriceAsc when the "Minimum Price" button is clicked', () => {
    const { getByTestId, getByText } = render(<Filter {...mockProps} />);

    const expandMoreButton = getByTestId('expand-price');
    fireEvent.click(expandMoreButton);

    const minimumPriceButton = getByText('Minimum Price');

    fireEvent.click(minimumPriceButton);
    expect(mockProps.onSortByPriceAsc).toHaveBeenCalled();
  });

  it('calls onSortByPriceDesc when the "Maximum Price" button is clicked', () => {
    const { getByTestId, getByText } = render(<Filter {...mockProps} />);

    const expandMoreButton = getByTestId('expand-price');
    fireEvent.click(expandMoreButton);

    const maximumPriceButton = getByText('Maximum Price');

    fireEvent.click(maximumPriceButton);
    expect(mockProps.onSortByPriceDesc).toHaveBeenCalled();
  });

  it('calls onCategoryChange when a category is clicked', () => {
    const { getByTestId, getByText } = render(<Filter {...mockProps} />);

    const expandMoreButton = getByTestId('id-category');
    fireEvent.click(expandMoreButton);

    const categoryButton = getByText('Chair');

    fireEvent.click(categoryButton);
    expect(mockProps.onCategoryChange).toHaveBeenCalled();
  });

  it('calls onSubCategoryChange when a category is clicked', () => {
    const { getByTestId, getByText, getAllByTestId, rerender } = render(<Filter {...mockProps} />);

    const expandMoreButton = getByTestId('id-category');
    fireEvent.click(expandMoreButton);
    const categoryButton = getByText('Chair');

    fireEvent.click(categoryButton);
    screen.debug();
    const expandSubCategory = getAllByTestId('expand-subcategory');
    fireEvent.click(expandSubCategory[1]);

    // Rerender to update the selected category because it is updated by function prop
    rerender(<Filter {...mockProps} selectedCategory="Chair" />);

    const subcategoryButton = getByText('Office chair');

    fireEvent.click(subcategoryButton);
    expect(mockProps.onCategoryChange).toHaveBeenCalled();
  });
});
