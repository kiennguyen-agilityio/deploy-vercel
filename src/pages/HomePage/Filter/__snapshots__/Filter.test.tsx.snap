// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="filter-container"
  >
    <h2
      class="filter-header"
    >
      Filter
    </h2>
    <div
      class="filter-section"
    >
      <h3
        class="filter-title "
      >
        Color
        <span
          aria-hidden="true"
          class="material-symbols-outlined "
          data-testid="expand-color"
        >
          expand_more
        </span>
      </h3>
    </div>
    <div
      class="filter-section"
    >
      <h3
        class="filter-title "
      >
        Rating
        <span
          aria-hidden="true"
          class="material-symbols-outlined "
        >
          expand_more
        </span>
      </h3>
    </div>
    <div
      class="filter-section"
    >
      <h3
        class="filter-title "
      >
        Price
        <span
          aria-hidden="true"
          class="material-symbols-outlined "
          data-testid="expand-price"
        >
          expand_more
        </span>
      </h3>
    </div>
    <div
      class="filter-section"
    >
      <h3
        aria-hidden="true"
        class="filter-title "
        data-testid="id-category"
      >
        Category
        <span
          class="material-symbols-outlined "
        >
          expand_more
        </span>
      </h3>
    </div>
  </div>
</DocumentFragment>
`;
